Improve unicode string handling in linkifier

jerch opened the pull request:
Had to do several changes to get the cell/length offset mess working. This PR shall fix #1669.

**Changes in `Buffer.ts`:**
- `BufferStringIterator`: class to get a hold of the buffer string content in a convenient way
- `Buffer.stringIndexToBufferIndex`: method to translate a string index from a string retrieved by `Buffer.translateBufferLineToString` to get the row and cols indices in the buffer. Atm this method assumes that the string started at col 0. ~(Can be changed easily to respect `startCol` as well.)~
- `Buffer.contents`: return a `BufferStringIterator`

**Changes in `Linkifier.ts`:**
- reimplemented `_linkifyRow` and `_doLinkifyRow`

**TODO:**

- [x] reimplement `Linkifier._addLink` to support unicode char offsets within a match (currently only pre/post match offsets are respected)
- [x] tests for unicode chars in linkifier

~@alexr00 The new methods in `Buffer` can also fix the highlight offset problems of the selection manager. Might be worth a look once this PR can land.~ ~_Seems to work now._~ issue created #1680 + #1686

jerch commented:
Ready for review :smile_cat:

Tyriar commented on src/Buffer.ts:
```diff
@@ -194,6 +194,39 @@ export class Buffer implements IBuffer {
+   * @param stringIndex index within the string
+   * @param startCol column offset the string was retrieved from
+   */
+  public stringIndexToBufferIndex(lineIndex: number, stringIndex: number): BufferIndex {
```
`SelectionManager._convertViewportColToCharacterIndex` and `SelectionManager._getWordAt` are closely related to this, not sure we can reuse any. Regardless, it's probably better not to do it in this PR.

jerch replied:
Yeah figured that. I think we should consolidate this in the future and remove doubled implementations from the code base, maybe with resolving #1686. Might also be a starting for your reflow proposal afterwards (its kinda interlinked).

Tyriar commented on src/CharWidth.ts:
```diff
@@ -169,3 +169,26 @@ export const wcwidth = (function(opts: {nul: number, control: number}): (ucs: nu
+
+/**
+ * Get the terminal cell width for a string.
+ * @param s
```
Better to exclude the @param if it doesn't have docs

jerch replied:
Done.

Tyriar commented on src/CharWidth.ts:
```diff
@@ -169,3 +169,26 @@ export const wcwidth = (function(opts: {nul: number, control: number}): (ucs: nu
+ * Get the terminal cell width for a string.
+ * @param s
+ */
+export function stringWidth(s: string): number {
```
This could do with some tests.

jerch replied:
Kinda hard to test, since it is interlinked with the char handling in `InputHandler.print`. Can try to compare the results with the sum of buffer line wcwidth values.
Imho this and `wcwidth` above needs a better integration by some string handling class in the future, that does all the nasty unicode and wcwidth stuff and can be used by `InputHandler` and `Buffer` methods at any time.

jerch replied:
Wrote some tests that compare the buffer wcwidths with the result of this function.

Tyriar commented on src/CharWidth.ts:
```diff
@@ -169,3 +169,26 @@ export const wcwidth = (function(opts: {nul: number, control: number}): (ucs: nu
+ * Get the terminal cell width for a string.
+ * @param s
+ */
+export function stringWidth(s: string): number {
```
`getStringCellWidth` is more explicit on what it does

jerch replied:
Done.

Tyriar commented on src/Linkifier.ts:
```diff
@@ -80,9 +81,24 @@ export class Linkifier extends EventEmitter implements ILinkifier {
-      this._linkifyRow(i);
+
+    const absoluteRowIndexStart = this._terminal.buffer.ydisp + this._rowsToLinkify.start;
+    if (absoluteRowIndexStart >= this._terminal.buffer.lines.length) {
```
Let's add the `// Ensure the row exists` back to help describe what's going on.

jerch replied:
Done.

Tyriar commented on src/Linkifier.ts:
```diff
@@ -153,99 +169,64 @@ export class Linkifier extends EventEmitter implements ILinkifier {
+      if (!uri) {
+        // something matched but does not comply with the given matchIndex
+        // since this is most likely a bug the regex itself we simply do nothing here
+        // TODO: should this be logged for debugging?
```
Maybe we should `throw new Error('...')` here so it's caught and fixed fast if it happens?

jerch replied:
Eww a hard `throw` here will also stop any later linkifier matchers, means if the first matcher regex is broken none will be linkified at all. Thats imho a good thing in a debug session yeah, but in live code it is likely to create tons of issue reports for any integrator. Do we have some debug flag setup I could hide this behind? Do we still use/propagate the `terminal.debug` flag?

jerch replied:
Ah you already created an issue for the debug  settings (#1560). Added the error behind terminal.debug for now.

Tyriar commented on src/Linkifier.ts:
```diff
@@ -153,99 +169,64 @@ export class Linkifier extends EventEmitter implements ILinkifier {
 
-    // Get index, match.index is for the outer match which includes negated chars
-    const index = text.indexOf(uri);
+      // due to complex regexes we cannot use match.index directly
```
I think my description on the left if a bit more clear, also we should remove the TODO as I don't think it's worth keeping in.

jerch replied:
Done.

jerch commented on src/Buffer.ts:
```diff
@@ -366,3 +403,38 @@ export class Marker extends EventEmitter implements IMarker {
+    return (withRanges) ? [range, result] : result;
+  }
+
+  public toArray(): string[] {
```
Yeah, removed it completely and replaced it with `next` invocations in the tests.

jerch commented on src/Buffer.ts:
```diff
@@ -366,3 +403,38 @@ export class Marker extends EventEmitter implements IMarker {
+  public toArray(): string[] {
+    const result = [];
+    while (this.hasNext()) {
+      result.push(this.next() as string);
```
ǹext`is currently used in he linkifier, I created an interface for the iterator result for now (always contains the range) and removed the overloaded return type.
I was at first unsure how to put/integrate this, even thought about a separate range only iterator. Also the whole iterator tries to solve some of the issues you addressed with your proposal for the reflow buffer redesign, imho it will see further redesign in the future.

jerch commented on src/Buffer.ts:
```diff
@@ -366,3 +403,38 @@ export class Marker extends EventEmitter implements IMarker {
+  private _end: number;
+  private _current: number;
+
+  constructor (private _buffer: IBuffer, private _trimRight: boolean, startIndex?: number, endIndex?: number) {
```
Done.

jerch commented on src/Linkifier.ts:
```diff
@@ -80,9 +81,24 @@ export class Linkifier extends EventEmitter implements ILinkifier {
+    // we skip those later in _doLinkifyRow
+    const linesIterator = this._terminal.buffer.contents(false, absoluteRowIndexStart, this._terminal.buffer.ydisp + this._rowsToLinkify.end + 1);
+    while (linesIterator.hasNext()) {
+      const lineData: any = linesIterator.next(true);
```
Must have been the cat :smile_cat: ....fixed.

jerch commented:
@Tyriar I think I have addressed all review issues. Up for the next round :smile_cat: 

Could you also test the issue from https://github.com/xtermjs/xterm.js/issues/1669#issuecomment-419718223 I think it is not related to the linkifier at all and will need separate tracking before closing the issue.
